src/transformers/testing_utils.py

```diff
@@ -1121,3 +1121,24 @@ def execute_subprocess_async(cmd, env=None, stdin=None, timeout=180, quiet=False
         raise RuntimeError(f"'{cmd_str}' produced no output.")
 
     return result
+
+
+def deep_round(obj, decimals=3):
+    """
+    Simplifies an object by rounding float numbers, and downcasting tensors/numpy arrays to get simple equality test
+    within tests.
+    """
+    from transformers.tokenization_utils import BatchEncoding
+
+    if isinstance(obj, list):
+        return [deep_round(item, decimals) for item in obj]
+    elif isinstance(obj, (dict, BatchEncoding)):
+        return {deep_round(k, decimals): deep_round(v, decimals) for k, v in obj.items()}
+    elif isinstance(obj, (str, int)):
+        return obj
+    elif is_torch_available() and isinstance(obj, torch.Tensor):
+        return deep_round(obj.tolist())
+    elif isinstance(obj, float):
+        return round(obj, decimals)
+    else:
+        raise Exception(f"Not supported: {type(obj)}")
```

Comment:
I'm not opposed to have this method! I think it could be made a bit better, for example with TF support. 

Also, wouldn't it be simpler to slightly change it so that it is a deep `torch.allclose`? We're used to having `torch.allclose` across the library, and we have some intuition of what `atol` or `rtol` we want. I think it would both stay closer to other tests, and make your testing code simpler, to have such a method.

I.e., instead of 

```py
self.assertEqual(
            deep_round(inputs),
            {
                # en_XX, A, test, EOS
                "input_ids": [[128022, 58, 4183, 2]],
                "attention_mask": [[1, 1, 1, 1]],
                # ar_AR
                "forced_bos_token_id": 44,
            },
        )
```

you would have

```py
self.assertTrue(
            deep_allclose(inputs, {
                # en_XX, A, test, EOS
                "input_ids": [[128022, 58, 4183, 2]],
                "attention_mask": [[1, 1, 1, 1]],
                # ar_AR
                "forced_bos_token_id": 44,
            }, atol=xxx, rtol=xxx)
        )
```

I don't feel strongly though, your current method (given TF support) is good already.

Reply:
The biggest difference between your version and my version is understanding the `diff` when there is a mismatch.

If input_ids is changed, but the rest of the entries are equal, then usually the unittest output is very readable.

assertTrue on the other hand is a pretty nasty "False is not true" message, which doesn't really help (and for larger structures it's sometimes hard to pinpoint where the difference happened.

Your solution is what I was referring to when mentionning the `assertAlmostEqual` option (where I think we need to fuse `assertTrue` and `deep_allclose` to get consistent error messages, yet be able to keep allclose like comparisons.

For this PR, I'll stick to just add TF support for `deep_round` as its much simpler and keeps readable diffs.

Reply:
Ok, this works for me. Thanks!